### PR TITLE
Added new priority_queue for ordering queded requests

### PR DESCRIPTION
**Pull request by gcatron**
Summary: This PR adds an optional priority field to runNetwork on HostManager. Also it adds a priority queue. 
When a request is made of the HostManager the following happens:
The request is added to the priorityQueue if the max queue size limit has been reached return an error immediately. 
If there are less active requests than the activeRequestLimit dispatchNextRun() is called.
If the activeLimit has been reached return.

Documentation: Updated in code documentation for new data structure and priority parameter in runNetwork.

Test Plan: Added a new QueueTest to HostManagerTest.

### CODE REVIEW

File: tests/unittests/HostManagerTest.cpp [hunk 1]
```diff
@@ -217,3 +218,60 @@ TEST_F(HostManagerTest, ConfigureHostManager) {
...
+  hostManager->runNetwork(
+      "main", std::move(context2),
+      [&run2p, &counter](RunIdentifierTy runID, llvm::Error err,
+                         std::unique_ptr<ExecutionContext> context) {
+        EXIT_ON_ERR(std::move(err));
+        run2p.set_value(counter++);
+      },
+      1);
+  // Set the priority of the second to 0 so it should be first in the queue
+  // aftter run1.
```
Comment (beicy) [inline tests/unittests/HostManagerTest.cpp]: nit : aftter->after

File: lib/Runtime/HostManager/HostManager.cpp [hunk 2]
```diff
@@ -269,10 +269,48 @@ llvm::Error HostManager::runNetworkBlocking(llvm::StringRef networkName,
...
+        {
+          std::lock_guard<std::mutex> networkLock(networkLock_);
+          if (inferQueue_.size()) {
+            // Get the next request, unfortunately priority_queue only provides
+            // a const ref to the top element, since we need to move it we first
+            // cast it to remove the const.
+            auto request =
+                std::move(const_cast<InferRequest &>(inferQueue_.top()));
+            inferQueue_.pop();
+            dispatchRun(std::move(request));
```
Comment (beicy) [inline lib/Runtime/HostManager/HostManager.cpp]: When a request is pop from the queue, do we need to add 1 to  activeRequestCount_? Or it will be added somewhere?

File: include/glow/Runtime/HostManager/HostManager.h [hunk 3]
```diff
@@ -50,15 +51,41 @@ class HostManager final {
     /// safety.
     std::atomic<size_t> refcount;
   };
+  /// Container for inference requests waiting in the queue.
+  struct InferRequest {
+    std::string networkName;
+    std::unique_ptr<ExecutionContext> context;
+    ResultCBTy callback;
+    uint64_t priority;
```
Comment (rdzhabarov) [inline include/glow/Runtime/HostManager/HostManager.h]: is priority going to be provided at onnxifi level?

File: lib/Runtime/HostManager/HostManager.cpp (hunk 2 above)
Comment (gcatron) [reply lib/Runtime/HostManager/HostManager.cpp]: activeRequest gets decremented if there are no more requests in the queue, but queuedRequestCount_ definitely needs to be decremented.

File: include/glow/Runtime/HostManager/HostManager.h (hunk 3 above)
Comment (gcatron) [reply include/glow/Runtime/HostManager/HostManager.h]: yes the plumbing for that still needs to be added. It should be straight forward so I'll add it to this PR.

File: include/glow/Runtime/HostManager/HostManager.h [hunk 4]
```diff
@@ -124,9 +154,11 @@ class HostManager final {
   /// Note: This method is intended to be thread-safe, it will be called
   /// concurrently from multiple threads.
   /// Returns -1 if networkName not found or too many active requests.
+  /// The parameter \p priority is used to indicate queuing priority, priotity
```
Comment (rdzhabarov) [inline include/glow/Runtime/HostManager/HostManager.h]: typo

**Comment by nickgg** [comment]
@gcatron I thought this was a feature request to address a specific thrashing problem where we increase latency by interleaving partial parts of a network rather than running that network to the end, does this change fix that? Can we have a test that demonstrates it does?

**Comment by gcatron** [comment]
@nickgg  This is addressing the thrash that sub-batching can introduce. where two requests get their sub-batches interleaved. Now if each request uses a different priority and assigned that priority to all of it's sub-batches they will be grouped together at execution. The QueueTest in HostManagerTest was added to test this functionality.

File: lib/Runtime/HostManager/HostManager.cpp [hunk 5]
```diff
@@ -344,37 +383,42 @@ HostManager::runNetwork(llvm::StringRef networkName,
...
   }
-
+  // Setup the request
+  InferRequest queuedRequest;
+  queuedRequest.networkName = networkName;
+  queuedRequest.context = std::move(context);
+  queuedRequest.priority = priority;
+  queuedRequest.callback = callback;
+  queuedRequest.requestID = currentRun;
+  // If we haven't reached maxActuveRequests kick off request.
```
Comment (bertmaher) [inline lib/Runtime/HostManager/HostManager.cpp]: typo: `maxActuveRequests`

File: lib/Runtime/HostManager/HostManager.cpp [hunk 6]
```diff
@@ -344,37 +383,42 @@ HostManager::runNetwork(llvm::StringRef networkName,
         std::move(context));
     return currentRun;
   }
-
+  // Setup the request
+  InferRequest queuedRequest;
```
Comment (bertmaher) [inline lib/Runtime/HostManager/HostManager.cpp]: Why not have a proper constructor for `InferRequest`, instead of initializing fields one-by-one?

**Comment by bertmaher** [comment]
I'll write up what I think I see in a moment, I just wanted to put a hold on this in case you were about to land it :-p

File: lib/Runtime/HostManager/HostManager.cpp [hunk 7]
```diff
@@ -344,37 +383,42 @@ HostManager::runNetwork(llvm::StringRef networkName,
...
-                     }
-                   }
-                   TRACE_EVENT_INSTANT(context->getTraceContext(),
-                                       TraceLevel::RUNTIME, "finish_" + name);
-                   callback(runID, std::move(err), std::move(context));
-                   --activeRequestCount_;
-                 });
+  {
+    std::lock_guard<std::mutex> networkLock(networkLock_);
+    inferQueue_.push(std::move(queuedRequest));
```
Comment (bertmaher) [inline lib/Runtime/HostManager/HostManager.cpp]: So I think it might be possible to drop requests where we're close to the active request limit.  Here's what I suspect can happen:
Thread 1: request1 comes in, less than maxActive, dispatchRun is called, submits r1 to executor (T3)
Thread 2: request2 comes in, greater than maxActive, skips dispatchRun, but before it can be enqueued...
Thread 3 (executor): processes request1, sees an empty queue, returns
Thread 2: enqueues r2
... and nothing ever calls dispatchRun to process r2.

In fact it seems like the service could deadlock, if a bunch of requests pile up between the executor and the enqueue.

The control flow and synchronization here is pretty confusing so maybe this can't happen... but can you convince me it's not possible?

**Comment by jackm321** [comment]
How do you envision priority being used?  I would think that a deadline might be more useful but I guess this could be used for that in a way

File: include/glow/Runtime/HostManager/HostManager.h [hunk 8]
```diff
@@ -50,15 +51,51 @@ class HostManager final {
...
+    }
+  };
 
   /// Count of current in-flight networks being run. Atomic to allow
   /// concurrency in runNetwork.
   std::atomic<size_t> activeRequestCount_{0};
 
+  /// Count of current in-queue networks. Atomic to allow
+  /// concurrency in runNetwork.
+  std::atomic<size_t> queuedRequestCount_{0};
```
Comment (jackm321) [inline include/glow/Runtime/HostManager/HostManager.h]: why not use inferQueue_.size()?

File: lib/Runtime/HostManager/HostManager.cpp (hunk 7 above)
Comment (gcatron) [reply lib/Runtime/HostManager/HostManager.cpp]: I think you're right, the second group of requests would be stuck until another request comes in, I'll fix it.

**Comment by gcatron** [comment]
> How do you envision priority being used? I would think that a deadline might be more useful but I guess this could be used for that in a way

Notionally this allows for sub-batches to be grouped together, the thought would be a timestamp of submission time but a timestamp of a deadline would work too. Earliest deadline would be run first.

File: lib/Runtime/HostManager/HostManager.cpp (hunk 7 above)
Comment (jackm321) [reply lib/Runtime/HostManager/HostManager.cpp]: I'm just skimming this, but that makes sense to me. Basically if thread2 gets descheduled at HostManager.cpp::399 while thread3 finishes the first request, nothing will pick up its request.
Also I agree that this is a bit confusing. The atomic state shared between the mutexed queues and the std::atomic counters describing the queues is a bit hard to follow.

Comment (gcatron) [reply lib/Runtime/HostManager/HostManager.cpp]: I think some of this complexity is I was trying to avoid enqueuing a request if it could be immediately dispatched. I think if I add the request to the queue first then there should be no race condition. 
T1 R1 comes in is enqueued, is dispatched
T2 R2, gets enqueued, if T1 finishes and grabs R2, then maxActive will still be at the limit, if T1 finishes early then maxActive will drop and R2 will get dispatched.
This means some amount of increased latency but I think that's better than more locking.

**Comment by jackm321** [comment]
> Notionally this allows for sub-batches to be grouped together, the thought would be a timestamp of submission time but a timestamp of a deadline would work too. Earliest deadline would be run first.

Ah ok I see, thanks. FWIW from previous experience, deadline is often more useful.

File: tests/unittests/HostManagerTest.cpp [hunk 9]
```diff
@@ -205,7 +206,7 @@ TEST_F(HostManagerTest, ConfigureHostManager) {
                           [lock](RunIdentifierTy runID, llvm::Error err,
                                  std::unique_ptr<ExecutionContext> context_) {
                             errToBool(std::move(err));
-                            std::unique_lock<std::mutex> guard(*lock);
+                            // std::unique_lock<std::mutex> guard(*lock);
```
Comment (bertmaher) [inline tests/unittests/HostManagerTest.cpp]: commented code?

Comment (gcatron) [reply tests/unittests/HostManagerTest.cpp]: removed now

File: lib/Runtime/HostManager/HostManager.cpp [hunk 10]
```diff
@@ -316,10 +316,45 @@ llvm::Error HostManager::runNetworkBlocking(llvm::StringRef networkName,
   return std::move(*DCHECK_NOTNULL(runErr.get()));
 }
 
+void HostManager::dispatchNextRun() {
+  {
```
Comment (bertmaher) [inline lib/Runtime/HostManager/HostManager.cpp]: This scope looks redundant

File: lib/Runtime/HostManager/HostManager.cpp [hunk 11]
```diff
@@ -316,10 +316,44 @@ llvm::Error HostManager::runNetworkBlocking(llvm::StringRef networkName,
...
+            std::lock_guard<std::mutex> networkLock(networkLock_);
+            auto it = networks_.find(name);
+            if (it != networks_.end()) {
+              it->second.refcount--;
+            }
+          }
+          TRACE_EVENT_INSTANT(context->getTraceContext(), TraceLevel::RUNTIME,
+                              "finish_" + name);
+          callback(runID, std::move(err), std::move(context));
+          dispatchNextRun();
```
Comment (bertmaher) [inline lib/Runtime/HostManager/HostManager.cpp]: The fact that we `dispatchNextRun` from inside a callback created by `dispatchNextRun` bugs me.  Is this, like, a common pattern in concurrent code?

It feels like what we really want is for the HostManager to push tasks into the priority queue, and to have a priority-aware ThreadPoolExecutor that knows how to pull the top pri item in its event loop.

cc @tracelogfb, just in case I'm out of my depth in concurrent program design...

File: lib/Runtime/HostManager/HostManager.cpp [hunk 12]
```diff
@@ -334,47 +368,43 @@ HostManager::runNetwork(llvm::StringRef networkName,
...
+          MAKE_ERR(
+              GlowErr::ErrorCode::RUNTIME_REQUEST_REFUSED,
+              strFormat(
+                  "The number of allowed queued requests has been exceeded. "
+                  "queued requests: %lu allowed requests: %zu",
+                  queueSize, config_.maxQueueSize)),
+          std::move(context));
+      return currentRun;
+    }
+    inferQueue_.push(std::move(queuedRequest));
```
Comment (rdzhabarov) [inline lib/Runtime/HostManager/HostManager.cpp]: does not it race with inferQueue_.pop() https://github.com/pytorch/glow/pull/3416/files#diff-0ff368c87ec74b6b37e29804c4122427R327 ? (It seems that we do push without any lock here).

Comment (bertmaher) [reply lib/Runtime/HostManager/HostManager.cpp]: There's a lock on L365 it's just waaaay up the page (hopefully I'm counting braces correctly :-p )

Comment (rdzhabarov) [reply lib/Runtime/HostManager/HostManager.cpp]: ah, right '}' was removed and expand that lock to larger scope.

Comment (gcatron) [reply lib/Runtime/HostManager/HostManager.cpp]: hmm yeah the push should be under a lock.

Comment (rdzhabarov) [reply lib/Runtime/HostManager/HostManager.cpp]: one more dumb comment while we are here :)

Invoking callback under the lock might be a risky business ...

Comment (bertmaher) [reply lib/Runtime/HostManager/HostManager.cpp]: That's an interesting point, given we don't know what the callback does, it may not be totally safe to hold the lock through it.  @gcatron can you follow up?
